atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownConcludedTransactionsTest.java

```diff
@@ -0,0 +1,164 @@
+        verify(knownConcludedTransactionsStore).get();
+    }
+
+    @Test
+    public void isKnownConcludedDoesNotPerformDatabaseReadIfAnswerAlreadyKnown() {
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(85L, Consistency.REMOTE_READ))
+                .isTrue();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(47L, Consistency.REMOTE_READ))
+                .isTrue();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(62L, Consistency.REMOTE_READ))
+                .isTrue();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(
+                        DEFAULT_RANGE.upperEndpoint(), Consistency.REMOTE_READ))
+                .isTrue();
+        verify(knownConcludedTransactionsStore).get();
+    }
+
+    @Test
+    public void isKnownConcludedAttemptsToLoadNewInformationIfAnswerNotCurrentlyKnown() {
+        // In each instance, it is possible that external factors have modified the database so that the transaction
+        // becomes known to have concluded, so a remote read is needed on each attempt.
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(2L, Consistency.REMOTE_READ))
+                .isFalse();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(2L, Consistency.REMOTE_READ))
+                .isFalse();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(2L, Consistency.REMOTE_READ))
+                .isFalse();
+        verify(knownConcludedTransactionsStore, times(3)).get();
+
+        when(knownConcludedTransactionsStore.get())
+                .thenReturn(Optional.of(TimestampRangeSet.singleRange(Range.closed(0L, 100L))));
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(2L, Consistency.REMOTE_READ))
+                .isTrue();
+        verify(knownConcludedTransactionsStore, times(4)).get();
+    }
+
+    @Test
+    public void addConcludedTimestampsSupplementsUnderlyingStore() {
+        defaultKnownConcludedTransactions.addConcludedTimestamps(ADDITIONAL_RANGE);
+        verify(knownConcludedTransactionsStore).supplement(ADDITIONAL_RANGE);
+    }
+
+    @Test
+    public void addConcludedTimestampsFlushesWritesToTheCache() {
+        defaultKnownConcludedTransactions.addConcludedTimestamps(ADDITIONAL_RANGE);
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(
+                        DEFAULT_RANGE.upperEndpoint() + 1, Consistency.LOCAL_READ))
+                .isTrue();
+        assertThat(defaultKnownConcludedTransactions.isKnownConcluded(
+                        DEFAULT_RANGE.upperEndpoint() + 1, Consistency.REMOTE_READ))
+                .isTrue();
+        verify(knownConcludedTransactionsStore, never()).get();
+    }
+
+    @Test
+    public void addConcludedTimestampsConcurrency() throws InterruptedException {
+        int numThreads = 10; // Concurrency should not be too high given this is accessed from background threads.
```

Comment:
this can as high as max(number of sweep threads, shards), right?

Reply:
Actually it's just |nodes|: the thread that updates these is the one that consumes the minimum sweep timestamp across all shards I think

Reply:
Yep that makes sense